File: experiment/harness/doctor.go
package harness

import (
	"fmt"
	"os/exec"
	"path/filepath"
	"strings"
)

func wordCount(value string) int { return len(strings.Fields(value)) }
func ratio(a, b int) float64 {
	difference := a - b
	if difference < 0 {
		difference = -difference
	}
	maximum := a
	if b > a {
		maximum = b
	}
	if maximum == 0 {
		return 0
	}
	return float64(difference) / float64(maximum)
}
func (h *Harness) checkWrappers() []string {
	errors := []string{}
	rows, err := loadSpecialists(h.Root)
	if err != nil {
		return []string{err.Error()}
	}
	for _, role := range allRoles {
		row := rows[role]
		functional, fictional := wordCount(row["functional_wrapper"]), wordCount(row["fictional_wrapper"])
		difference := ratio(functional, fictional)
		fmt.Printf("wrapper %-13s functional=%2d fictional=%2d difference=%.1f%%\n", role, functional, fictional, difference*100)
		if difference > .10 {
			errors = append(errors, role+" wrappers differ by more than 10% in whitespace-token count")
		}
	}
	var wrappers map[string]string
	if err := readJSON(filepath.Join(h.Root, "experiment/prompts/omnibus-wrappers.json"), &wrappers); err != nil {
		return append(errors, err.Error())
	}
	functional, fictional := wordCount(wrappers["functional"]), wordCount(wrappers["fictional"])
	difference := ratio(functional, fictional)
	fmt.Printf("wrapper %-13s functional=%2d fictional=%2d difference=%.1f%%\n", "omnibus", functional, fictional, difference*100)
	if difference > .10 {
		errors = append(errors, "omnibus wrappers differ by more than 10%")
	}
	return errors
}
func (h *Harness) checkPrompts(config Config) ([]string, Plan) {
	errors := []string{}
	plan, err := BuildPlan(config)
	if err != nil {
		return []string{err.Error()}, Plan{}
	}
	callMap := map[string]Call{}
	for _, call := range plan.Calls {
		callMap[call.CallID] = call
	}
	for _, call := range plan.Calls {
		kind := stringValue(call.PromptSpec["kind"])
		var prompt string
		var err error
		if kind == "generic" || kind == "omnibus" || kind == "specialist" {
			prompt, err = staticPrompt(h.Root, call)
		} else if kind == "fuser" || kind == "chain" {
			deps := []dependencyPair{}
			for _, id := range call.DependsOn {
				deps = append(deps, dependencyPair{Call: callMap[id], Response: map[string]any{"findings": []any{}}})
			}
			prompt, err = dynamicPrompt(h.Root, call, deps)
		} else {
			err = fmt.Errorf("unknown prompt kind %s", kind)
		}
		if err == nil {
			var answer string
			answer, err = answerText(h.Root, call.Packet)
			if err == nil {
				err = assertAnswerKeyAbsent(prompt, answer)
			}
		}
		if err != nil {
			errors = append(errors, fmt.Sprintf("prompt %s: %v", call.CallID, err))
		}
	}
	fmt.Printf("assembled and contamination-checked %d planned prompts\n", len(plan.Calls))
	fmt.Printf("plan contains %d calls and %d scoreable output sets\n", len(plan.Calls), len(plan.OutputSets))
	return errors, plan
}
func runLookup(name string, args ...string) (string, error) {
	cmd := exec.Command(name, args...)
	out, err := cmd.CombinedOutput()
	return string(out), err
}
func (h *Harness) checkModels(config Config, skip bool) []string {
	if skip {
		return nil
	}
	providers := config.Providers
	if len(providers) == 0 {
		providers = []Provider{config.Provider}
	}
	errors := []string{}
	for _, p := range providers {
		if p.Adapter == "mock" {
			continue
		}
		if p.Adapter == "agy" {
			errors = append(errors, "agy is disabled: OAuth/keychain state conflicts with strict ephemeral-home isolation; use Pi with an explicit Google model")
			continue
		}
		if _, err := exec.LookPath(p.Adapter); err != nil {
			errors = append(errors, "missing executable: "+p.Adapter)
			continue
		}
		switch p.Adapter {
		case "pi", "claude":
			out, err := runLookup("pi", "--no-extensions", "--list-models", p.Model)
			registry := p.Model
			if parts := strings.SplitN(registry, "/", 2); len(parts) == 2 {
				registry = parts[1]
			}
			if err != nil || !strings.Contains(out, registry) {
				errors = append(errors, "model is not in local Pi registry: "+p.Model)
			}
		default:
			errors = append(errors, "unknown adapter: "+p.Adapter)
		}
	}
	return errors
}
func (h *Harness) Doctor(configPath string, skipLookup bool) error {
	if err := currentUserCheck(); err != nil {
		return err
	}
	if err := h.SandboxProbe(); err != nil {
		return err
	}
	if !filepath.IsAbs(configPath) {
		configPath = filepath.Join(h.Root, configPath)
	}
	var config Config
	if err := readJSON(configPath, &config); err != nil {
		return err
	}
	errors := h.checkWrappers()
	promptErrors, _ := h.checkPrompts(config)
	errors = append(errors, promptErrors...)
	errors = append(errors, h.checkModels(config, skipLookup)...)
	if len(errors) > 0 {
		fmt.Println("Doctor failed:")
		for _, err := range errors {
			fmt.Println("- " + err)
		}
		return fmt.Errorf("doctor found %d error(s)", len(errors))
	}
	fmt.Println("Doctor passed. No model calls were made.")
	return nil
}

File: experiment/harness/doctor.py
from __future__ import annotations

import argparse
import json
from pathlib import Path
import shutil
import subprocess

from .common import ROOT, load_json
from .plan import build
from .prompting import answer_text, assert_answer_key_absent, dynamic_prompt, static_prompt


def words(value: str) -> int:
    return len(value.split())


def ratio(a: int, b: int) -> float:
    return abs(a - b) / max(a, b)


def check_wrappers() -> list[str]:
    errors: list[str] = []
    specialists = load_json(ROOT / "experiment/prompts/specialists.json")
    for row in specialists:
        functional = words(row["functional_wrapper"])
        fictional = words(row["fictional_wrapper"])
        difference = ratio(functional, fictional)
        print(
            f"wrapper {row['id']:<13} functional={functional:2d} fictional={fictional:2d} difference={difference:.1%}"
        )
        if difference > 0.10:
            errors.append(f"{row['id']} wrappers differ by more than 10% in whitespace-token count")
    wrappers = load_json(ROOT / "experiment/prompts/omnibus-wrappers.json")
    functional, fictional = words(wrappers["functional"]), words(wrappers["fictional"])
    difference = ratio(functional, fictional)
    print(f"wrapper {'omnibus':<13} functional={functional:2d} fictional={fictional:2d} difference={difference:.1%}")
    if difference > 0.10:
        errors.append("omnibus wrappers differ by more than 10% in whitespace-token count")
    return errors


def lookup_models(config: dict, skip: bool) -> list[str]:
    errors: list[str] = []
    providers = config.get("providers", [config.get("provider")])
    for provider in [value for value in providers if value]:
        adapter, model = provider["adapter"], provider["model"]
        executable = {"claude": "claude", "agy": "agy", "pi": "pi", "mock": None}[adapter]
        if executable and not shutil.which(executable):
            errors.append(f"missing executable: {executable}")
            continue
        if skip or adapter == "mock":
            continue
        if adapter == "agy":
            result = subprocess.run(["agy", "models"], text=True, capture_output=True, timeout=60)
            if result.returncode or model not in result.stdout:
                errors.append(f"agy model is not listed: {model}")
        elif adapter in {"pi", "claude"}:
            result = subprocess.run(
                ["pi", "--no-extensions", "--list-models", model],
                text=True,
                capture_output=True,
                timeout=60,
            )
            registry_id = model.split("/", 1)[-1]
            if result.returncode or registry_id not in result.stdout:
                errors.append(f"model is not in the local pi registry: {model}")
    return errors


def check_prompts(config: dict) -> list[str]:
    errors: list[str] = []
    plan = build(config)
    call_map = {call["call_id"]: call for call in plan["calls"]}
    for call in plan["calls"]:
        kind = call["prompt_spec"]["kind"]
        try:
            if kind in {"generic", "omnibus", "specialist"}:
                prompt = static_prompt(ROOT, call)
            elif kind == "fuser":
                dependencies = [
                    (call_map[call_id], {"findings": []}) for call_id in call["depends_on"]
                ]
                prompt = dynamic_prompt(ROOT, call, dependencies)
            elif kind == "chain":
                dependency = call_map[call["depends_on"][0]]
                prompt = dynamic_prompt(ROOT, call, [(dependency, {"findings": []})])
            else:
                errors.append(f"unknown prompt kind: {kind}")
                continue
            assert_answer_key_absent(prompt, answer_text(ROOT, call["packet"]))
        except Exception as error:  # surfaced with call identity for diagnosis
            errors.append(f"prompt {call['call_id']}: {error}")
    print(f"assembled and contamination-checked {len(plan['calls'])} planned prompts")
    print(f"plan contains {len(plan['calls'])} calls and {len(plan['output_sets'])} scoreable output sets")
    return errors


def main() -> int:
    parser = argparse.ArgumentParser(description="Validate an experiment config without model calls")
    parser.add_argument("config")
    parser.add_argument("--skip-model-lookup", action="store_true")
    args = parser.parse_args()
    config_path = Path(args.config)
    if not config_path.is_absolute():
        config_path = ROOT / config_path
    config = load_json(config_path)

    errors = []
    errors.extend(check_wrappers())
    errors.extend(check_prompts(config))
    errors.extend(lookup_models(config, args.skip_model_lookup))
    if errors:
        print("Doctor failed:")
        for error in errors:
            print(f"- {error}")
        return 1
    print("Doctor passed. No model calls were made.")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
